### PR TITLE
Make String and box types frozen by default, more freeze checks

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 1]
```diff
@@ -1397,27 +1398,33 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
             assert (value.receiver == null)
             val ptr = context.llvmDeclarations.forStaticField(value.symbol.owner).storage
             return functionGenerationContext.loadSlot(ptr, value.descriptor.isVar())
         }
     }
 
     //-------------------------------------------------------------------------//
+    private fun needMutationCheck(expression: IrExpression): Boolean {
+        // For now we omit mutation checks on immutable types, as this allows initialization in constructor
+        // and it is assumed API doesn't allow to change them.
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: > it is

its?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 2]
```diff
@@ -1397,27 +1398,33 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
 
     private fun evaluateSetField(value: IrSetField): LLVMValueRef {
         context.log{"evaluateSetField               : ${ir2string(value)}"}
         val valueToAssign = evaluateExpression(value.value)
         if (value.descriptor.dispatchReceiverParameter != null) {
             val thisPtr = evaluateExpression(value.receiver!!)
-            functionGenerationContext.call(context.llvm.mutationCheck,
-                    listOf(functionGenerationContext.bitcast(codegen.kObjHeaderPtr, thisPtr)),
-                    Lifetime.IRRELEVANT, ExceptionHandler.Caller)
+            if (needMutationCheck(value.receiver!!)) {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: The check based on receiver type doesn't look reliable to me. We can get `ClassDescriptor` from field descriptor instead, e.g.`value.descriptor.containingDeclaration`.

File: runtime/src/main/cpp/Arrays.cpp [hunk 3]
```diff
@@ -379,6 +404,7 @@ KFloat Kotlin_FloatArray_get(KConstRef thiz, KInt index) {
   if (static_cast<uint32_t>(index) >= array->count_) {
     ThrowArrayIndexOutOfBoundsException();
   }
+  mutabilityCheck(thiz);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Arrays.cpp]: Mutability check in array `get`?

File: runtime/src/main/kotlin/konan/Annotations.kt [hunk 4]
```diff
@@ -61,6 +61,13 @@ public annotation class Used
 @Retention(AnnotationRetention.SOURCE)
 public annotation class Throws(vararg val exceptionClasses: KClass<out Throwable>)
 
+/**
+ * Class is immutable and is frozen by default.
+ */
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
+annotation class Immutable
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/Annotations.kt]: This annotation is unchecked and thus unsafe. Shouldn't it be internal yet?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 2 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: It does not have `findAnnotation`, it seems.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 5]
```diff
@@ -1397,27 +1398,33 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
...
-        else {
+        } else {
             assert (value.receiver == null)
             val ptr = context.llvmDeclarations.forStaticField(value.symbol.owner).storage
             return functionGenerationContext.loadSlot(ptr, value.descriptor.isVar())
         }
     }
 
     //-------------------------------------------------------------------------//
+    private fun needMutationCheck(expression: IrExpression): Boolean {
```
[the remark this answers is not in the document]
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Well, with object fields it got somewhat complicated, so let's keep this change simple.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 1 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Hmm, not really.

File: runtime/src/main/cpp/Arrays.cpp (hunk 3 above)
Comment (olonho) [reply runtime/src/main/cpp/Arrays.cpp]: Mutability shall only be tested in setter, isn't it?

File: runtime/src/main/kotlin/konan/Annotations.kt (hunk 4 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/Annotations.kt]: OK

File: runtime/src/main/cpp/Arrays.cpp (hunk 3 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Arrays.cpp]: Exactly, and this function is named `Kotlin_FloatArray_get`. Is it ok?

File: backend.native/tests/runtime/workers/freeze2.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+import kotlin.test.*
+
+import konan.worker.*
+
+data class Data(var int: Int)
+
+/* An awful hack: we mark data as immutable, while it is not. */
+@konan.internal.Immutable
+data class SharedData(var field: Any)
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/freeze2.kt]: What is the motivation behind adding test for an awful hack?

File: backend.native/tests/runtime/workers/freeze2.kt [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    ok = false
+    val a1 = ByteArray(2)
+    a1[1]++
+    a1.freeze()
+    try {
+        a1[1]++
+    } catch (e: InvalidMutabilityException) {
+        ok = true
+    }
+    if (!ok) throw Error("can mutate frozen ByteArray")
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/freeze2.kt]: We have some `kotlin.test.*` methods to be used in this case, e.g. [`assertFails`](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/assert-fails.html) and [`assertFailsWith`](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/assert-fails-with.html).

File: backend.native/tests/runtime/workers/freeze2.kt [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    // Ensure that String and integral boxes are frozen by default.
+    val worker = startWorker()
+    val data = SharedData("Hello")
+    worker.schedule(TransferMode.CHECKED, { data } ) {
+        input -> println("Worker 1: $input")
+    }.result()
+
+    data.field = 42
+    worker.schedule(TransferMode.CHECKED, { data } ) {
+        input -> println("Worker2: $input")
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/freeze2.kt]: How does this check that `Int` box is frozen by default?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 1 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: > and it is assumed API

Then "that" is missing between "assumed" and "API".

File: backend.native/tests/runtime/workers/freeze2.kt (hunk 7 above)
Comment (olonho) [reply backend.native/tests/runtime/workers/freeze2.kt]: Done

File: runtime/src/main/cpp/Arrays.cpp (hunk 3 above)
Comment (olonho) [reply runtime/src/main/cpp/Arrays.cpp]: OK

**Pull request merged**